packages/vite/src/node/server/index.ts

```diff
@@ -499,7 +501,16 @@ export async function _createServer(
   }
 
   for (const environment of Object.values(environments)) {
-    await environment.init({ watcher })
+    const previousEnvironment = options.previousEnvironments?.[environment.name]
+
+    await environment.init({
+      watcher,
+      previousInstance:
+        // only pass if the previous environment is an instance of the same class
+        previousEnvironment instanceof environment.constructor
+          ? previousEnvironment
+          : undefined,
```

Comment:
We were discussing with @sheremet-va moving away from classes and let people use plain objects if they want. This checks is needed for this kind of pattern though. If we use an interface, should we add a new `type: string` field that each environment provider can use? So they can define it as "workerd" for example. We would need to also avoid passing the previousInstance when the environment factory library is updated though (and the type name will remain the same in that case).

Maybe we need to keep the classes?

Reply:
> If we use an interface, should we add a new `type: string` field that each environment provider can use?

I think yes. It seems some of them already pass in an information like that. I'm not sure about the intent though.
https://github.com/flarelabs-net/vite-environment-providers/blob/21bbc0bf49b8267693b8ea76ad05380f02e7d2db/packages/vite-environment-provider-cloudflare/src/index.ts#L45-L50

Another way would be to just pass the `previousEnvironment` always and let the filter done by environment side. But I think that is a bit cumbersome for env providers.

Reply:
In the team meeting, we discussed about this and concluded that this won't be a reason to add a property. But if there's a need to add a property in other places, we can use that here.